source/algorithms.tex

```diff
@@ -10755,6 +10767,123 @@
 where the result of the division is truncated towards zero.
 \end{itemdescr}
 
+\rSec2[numeric.sat]{Saturation arithmetic}
+
+\rSec3[numeric.sat.func]{Arithmetic functions}
+
+\pnum
+\begin{note}
+In the following descriptions, an arithmetic operation
+is performed as a mathematical operation with infinite range and then
+it is determined whether the mathematical result fits into the result type.
+\end{note}
+
+\indexlibraryglobal{add_sat}%
+\begin{itemdecl}
+template<class T>
+  constexpr T add_sat(T x, T y) noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{T} is a signed or unsigned integer type\iref{basic.fundamental}.
+
+\pnum
+\returns
+If \tcode{x + y} is representable as a value of type \tcode{T}, \tcode{x + y};
+otherwise, either the largest or smallest representable value of type \tcode{T},
+% FIXME: Do we define what is means for one value to be "closer" to another?
+whichever is closer to the value of \tcode{x + y}.
+\end{itemdescr}
+
+\indexlibraryglobal{sub_sat}%
+\begin{itemdecl}
+template<class T>
+  constexpr T sub_sat(T x, T y) noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{T} is a signed or unsigned integer type\iref{basic.fundamental}.
+
+\pnum
+\returns
+If \tcode{x - y} is representable as a value of type \tcode{T}, \tcode{x - y};
+otherwise, either the largest or smallest representable value of type \tcode{T},
+whichever is closer to the value of \tcode{x - y}.
+\end{itemdescr}
+
+\indexlibraryglobal{mul_sat}%
+\begin{itemdecl}
+template<class T>
+  constexpr T mul_sat(T x, T y) noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{T} is a signed or unsigned integer type\iref{basic.fundamental}.
+
+\pnum
+\returns
+If \tcode{x * y} is representable as a value of type \tcode{T}, \tcode{x * y};
+otherwise, either the largest or smallest representable value of type \tcode{T},
+whichever is closer to the value of \tcode{x * y}.
+\end{itemdescr}
+
+\indexlibraryglobal{div_sat}%
+\begin{itemdecl}
+template<class T>
+  constexpr T div_sat(T x, T y) noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\tcode{T} is a signed or unsigned integer type\iref{basic.fundamental}.
+
+\pnum
+\expects
+\tcode{y != 0} is \tcode{true}.
+
+\pnum
+Let \tcode{q} be \tcode{x / y}, with any fractional part discarded.
+
+\pnum
+\returns
+If \tcode{q} is representable as a value of type \tcode{T}, \tcode{q};
+otherwise, either the largest or smallest representable value of type \tcode{T},
+whichever is closer to the value of \tcode{q}.
```

Comment:
It seems that we can simply the specification, since IIUC saturation is only needed when `x` is the most negative value and `y == -1`,
```suggestion
\pnum
\returns
If \tcode{T} is a signed integer type and \tcode{x == numeric_limits<T>::min() && y == -1}
is \tcode{true}, \tcode{numeric_limits<T>::max()}, otherwise, \tcode{x / y}.
```

Reply:
That's nice, done!